File: app/src/main/java/com/gigabytedevelopersinc/app/cometOTP/Preferences/CredentialsPreference.java
package com.gigabytedevelopersinc.app.cometOTP.Preferences;

import android.app.AlertDialog;
import android.app.KeyguardManager;
import android.content.Context;
import android.preference.DialogPreference;

import com.gigabytedevelopersinc.app.cometOTP.Utilities.GeneralUtils;
import com.google.android.material.textfield.TextInputEditText;
import com.google.android.material.textfield.TextInputLayout;
import android.text.Editable;
import android.text.InputType;
import android.text.TextWatcher;
import android.text.method.PasswordTransformationMethod;
import android.util.AttributeSet;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.ListView;
import android.widget.TextView;
import android.widget.Toast;

import com.gigabytedevelopersinc.app.cometOTP.R;
import com.gigabytedevelopersinc.app.cometOTP.Utilities.Constants;
import com.gigabytedevelopersinc.app.cometOTP.Utilities.Settings;
import com.gigabytedevelopersinc.app.cometOTP.Utilities.UIHelper;

import java.util.Arrays;
import java.util.List;
import java.util.Locale;
import java.util.Objects;

import static android.content.Context.KEYGUARD_SERVICE;
import static com.gigabytedevelopersinc.app.cometOTP.Utilities.Constants.AuthMethod;
import static com.gigabytedevelopersinc.app.cometOTP.Utilities.Constants.EncryptionType;

public class CredentialsPreference extends DialogPreference
    implements AdapterView.OnItemClickListener, View.OnClickListener, TextWatcher {
    public static final AuthMethod DEFAULT_VALUE = AuthMethod.NONE;

    public interface EncryptionChangeCallback {
        boolean testEncryptionChange(byte[] newKey);
    }

    private List<String> entries;
    private static final List<AuthMethod> entryValues = Arrays.asList(
            AuthMethod.NONE,
            AuthMethod.PASSWORD,
            AuthMethod.PIN,
            AuthMethod.DEVICE
    );

    private int minLength = 0;

    private Settings settings;
    private AuthMethod value = AuthMethod.NONE;
    private EncryptionChangeCallback encryptionChangeCallback = null;

    private LinearLayout credentialsLayout;
    private TextInputLayout passwordLayout;
    private TextInputEditText passwordInput;
    private EditText passwordConfirm;
    private TextView tooShortWarning;

    private Button btnSave;

    public CredentialsPreference(Context context, AttributeSet attrs) {
        super(context, attrs);

        settings = new Settings(context);
        entries = Arrays.asList(context.getResources().getStringArray(R.array.settings_entries_auth));

        setDialogLayoutResource(R.layout.component_authentication);
    }

    public void setEncryptionChangeCallback(EncryptionChangeCallback cb) {
        this.encryptionChangeCallback = cb;
    }

    @Override
    protected void onPrepareDialogBuilder(AlertDialog.Builder builder) {
        super.onPrepareDialogBuilder(builder);

        builder.setPositiveButton(null, null);
        builder.setNegativeButton(null, null);
        builder.setCancelable(false);
    }

    @Override
    protected void onBindDialogView(View view) {
        value = settings.getAuthMethod();

        ListView listView = view.findViewById(R.id.credentialSelection);

        ArrayAdapter<String> adapter = new ArrayAdapter<>(getContext(), android.R.layout.simple_list_item_single_choice, entries);
        listView.setAdapter(adapter);

        int index = entryValues.indexOf(value);
        listView.setSelection(index);
        listView.setItemChecked(index,true);
        listView.setOnItemClickListener(this);

        credentialsLayout = view.findViewById(R.id.credentialsLayout);

        passwordLayout = view.findViewById(R.id.passwordLayout);
        passwordInput = view.findViewById(R.id.passwordEdit);
        passwordConfirm = view.findViewById(R.id.passwordConfirm);

        if (settings.getBlockAccessibility()) {
            passwordLayout.setImportantForAccessibility(View.IMPORTANT_FOR_ACCESSIBILITY_NO_HIDE_DESCENDANTS);
            passwordConfirm.setImportantForAccessibility(View.IMPORTANT_FOR_ACCESSIBILITY_NO);
        }

        if (GeneralUtils.INSTANCE.isOreo() && settings.getBlockAutofill()) {
            passwordLayout.setImportantForAutofill(View.IMPORTANT_FOR_AUTOFILL_NO_EXCLUDE_DESCENDANTS);
            passwordConfirm.setImportantForAutofill(View.IMPORTANT_FOR_AUTOFILL_NO);
        }

        tooShortWarning = view.findViewById(R.id.tooShortWarning);

        passwordInput.addTextChangedListener(this);
        passwordConfirm.addTextChangedListener(this);

        Button btnCancel = view.findViewById(R.id.btnCancel);
        btnSave = view.findViewById(R.id.btnSave);

        btnCancel.setOnClickListener(this);
        btnSave.setOnClickListener(this);

        updateLayout();

        super.onBindDialogView(view);
    }

    @Override
    protected void onSetInitialValue(boolean restorePersistedValue, Object defaultValue) {
        if (restorePersistedValue) {
            String stringValue = getPersistedString(DEFAULT_VALUE.name().toLowerCase(Locale.ENGLISH));
            value = AuthMethod.valueOf(stringValue.toUpperCase(Locale.ENGLISH));
        } else {
            value = DEFAULT_VALUE;
            persistString(value.name().toLowerCase(Locale.ENGLISH));
        }

        setSummary(entries.get(entryValues.indexOf(value)));
    }

    private void saveValues() {
        byte[] newKey = null;

        if (settings.getEncryption() == EncryptionType.PASSWORD) {
            if (value == AuthMethod.NONE || value == AuthMethod.DEVICE) {
                UIHelper.showGenericDialog(getContext(), R.string.settings_dialog_title_error, R.string.settings_dialog_msg_auth_invalid_with_encryption);
                return;
            }
        }

        if (value == AuthMethod.DEVICE) {
            KeyguardManager km = (KeyguardManager) getContext().getSystemService(KEYGUARD_SERVICE);

            assert km != null;
            if (! km.isKeyguardSecure()) {
                Toast.makeText(getContext(), R.string.settings_toast_auth_device_not_secure, Toast.LENGTH_LONG).show();
                return;
            }
        }

        if (value == AuthMethod.PASSWORD || value == AuthMethod.PIN) {
            String password = Objects.requireNonNull(passwordInput.getText()).toString();
            if (!password.isEmpty()) {
                newKey = settings.setAuthCredentials(password);
            } else {
                return;
            }
        }

        if (settings.getEncryption() == EncryptionType.PASSWORD) {
            if (newKey == null || encryptionChangeCallback == null)
                return;

            if (! encryptionChangeCallback.testEncryptionChange(newKey))
                return;
        }

        persistString(value.toString().toLowerCase());
        setSummary(entries.get(entryValues.indexOf(value)));
    }

    @Override
    public void onClick(View view) {
        switch (view.getId()) {
            case (R.id.btnCancel):
                getDialog().dismiss();
                break;
            case (R.id.btnSave):
                saveValues();
                getDialog().dismiss();
                break;
            default:
                break;
        }
    }

    @Override
    public void onTextChanged(CharSequence s, int start, int before, int count) {
        String password = passwordInput.getEditableText().toString();

        if (password.length() >= minLength) {
            tooShortWarning.setVisibility(View.GONE);

            String confirm = passwordConfirm.getEditableText().toString();

            if (!password.isEmpty() && !confirm.isEmpty() && password.equals(confirm)) {
                btnSave.setEnabled(true);
            } else {
                btnSave.setEnabled(false);
            }
        } else {
            tooShortWarning.setVisibility(View.VISIBLE);
        }
    }

    private void updateLayout() {
        if (value == AuthMethod.NONE) {
            credentialsLayout.setVisibility(View.GONE);

            if (getDialog() != null)
                UIHelper.hideKeyboard(getContext(), getDialog().getCurrentFocus());

            btnSave.setEnabled(true);
        } else if (value == AuthMethod.PASSWORD) {
            credentialsLayout.setVisibility(View.VISIBLE);

            passwordLayout.setHint(getContext().getString(R.string.settings_hint_password));
            passwordConfirm.setHint(R.string.settings_hint_password_confirm);

            passwordInput.setInputType(InputType.TYPE_CLASS_TEXT | InputType.TYPE_TEXT_VARIATION_PASSWORD);
            passwordConfirm.setInputType(InputType.TYPE_CLASS_TEXT | InputType.TYPE_TEXT_VARIATION_PASSWORD);

            passwordInput.setTransformationMethod(new PasswordTransformationMethod());
            passwordConfirm.setTransformationMethod(new PasswordTransformationMethod());

            minLength = Constants.AUTH_MIN_PASSWORD_LENGTH;
            tooShortWarning.setText(getContext().getString(R.string.settings_label_short_password, minLength));

            passwordInput.requestFocus();
            UIHelper.showKeyboard(getContext(), passwordInput);

            btnSave.setEnabled(false);
        } else if (value == AuthMethod.PIN) {
            credentialsLayout.setVisibility(View.VISIBLE);

            passwordLayout.setHint(getContext().getString(R.string.settings_hint_pin));
            passwordConfirm.setHint(R.string.settings_hint_pin_confirm);

            passwordInput.setInputType(InputType.TYPE_CLASS_NUMBER | InputType.TYPE_NUMBER_VARIATION_PASSWORD);
            passwordConfirm.setInputType(InputType.TYPE_CLASS_NUMBER | InputType.TYPE_NUMBER_VARIATION_PASSWORD);

            passwordInput.setTransformationMethod(new PasswordTransformationMethod());
            passwordConfirm.setTransformationMethod(new PasswordTransformationMethod());

            minLength = Constants.AUTH_MIN_PIN_LENGTH;
            tooShortWarning.setText(getContext().getString(R.string.settings_label_short_pin, minLength));

            passwordInput.requestFocus();
            UIHelper.showKeyboard(getContext(), passwordInput);

            btnSave.setEnabled(false);
        } else if (value == AuthMethod.DEVICE) {
            credentialsLayout.setVisibility(View.GONE);

            if (getDialog() != null)
                UIHelper.hideKeyboard(getContext(), getDialog().getCurrentFocus());

            btnSave.setEnabled(true);
        }
    }

    @Override
    public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
        value = entryValues.get(position);
        updateLayout();
    }

    // Needed stub functions
    @Override
    public void afterTextChanged(Editable s) {}

    @Override
    public void beforeTextChanged(CharSequence s, int start, int count, int after) {}


}

File: app/src/main/java/com/gigabytedevelopersinc/app/cometOTP/Preferences/CredentialsPreference.kt
@file:Suppress("PackageName", "DEPRECATION", "OVERRIDE_DEPRECATION")
package com.gigabytedevelopersinc.app.cometOTP.Preferences

import android.app.AlertDialog
import android.app.KeyguardManager
import android.content.Context
import android.content.Context.KEYGUARD_SERVICE
import android.preference.DialogPreference
import android.text.Editable
import android.text.InputType
import android.text.TextWatcher
import android.text.method.PasswordTransformationMethod
import android.util.AttributeSet
import android.view.View
import android.widget.AdapterView
import android.widget.ArrayAdapter
import android.widget.Button
import android.widget.EditText
import android.widget.LinearLayout
import android.widget.ListView
import android.widget.TextView
import android.widget.Toast
import com.gigabytedevelopersinc.app.cometOTP.R
import com.gigabytedevelopersinc.app.cometOTP.Utilities.Constants
import com.gigabytedevelopersinc.app.cometOTP.Utilities.Constants.AuthMethod
import com.gigabytedevelopersinc.app.cometOTP.Utilities.Constants.EncryptionType
import com.gigabytedevelopersinc.app.cometOTP.Utilities.GeneralUtils
import com.gigabytedevelopersinc.app.cometOTP.Utilities.Settings
import com.gigabytedevelopersinc.app.cometOTP.Utilities.UIHelper
import com.google.android.material.textfield.TextInputEditText
import com.google.android.material.textfield.TextInputLayout
import java.util.Locale

class CredentialsPreference(context: Context, attrs: AttributeSet?) : DialogPreference(context, attrs),
    AdapterView.OnItemClickListener, View.OnClickListener, TextWatcher {

    fun interface EncryptionChangeCallback {
        fun testEncryptionChange(newKey: ByteArray): Boolean
    }

    private val entries: List<String>

    private var minLength = 0

    private val settings: Settings
    private var value = AuthMethod.NONE
    private var encryptionChangeCallback: EncryptionChangeCallback? = null

    private lateinit var credentialsLayout: LinearLayout
    private lateinit var passwordLayout: TextInputLayout
    private lateinit var passwordInput: TextInputEditText
    private lateinit var passwordConfirm: EditText
    private lateinit var tooShortWarning: TextView

    private lateinit var btnSave: Button

    init {
        settings = Settings(context)
        entries = context.resources.getStringArray(R.array.settings_entries_auth).asList()

        dialogLayoutResource = R.layout.component_authentication
    }

    fun setEncryptionChangeCallback(cb: EncryptionChangeCallback?) {
        this.encryptionChangeCallback = cb
    }

    override fun onPrepareDialogBuilder(builder: AlertDialog.Builder) {
        super.onPrepareDialogBuilder(builder)

        builder.setPositiveButton(null, null)
        builder.setNegativeButton(null, null)
        builder.setCancelable(false)
    }

    override fun onBindDialogView(view: View) {
        value = settings.authMethod

        val listView = view.findViewById<ListView>(R.id.credentialSelection)

        val adapter = ArrayAdapter(context, android.R.layout.simple_list_item_single_choice, entries)
        listView.adapter = adapter

        val index = entryValues.indexOf(value)
        listView.setSelection(index)
        listView.setItemChecked(index, true)
        listView.onItemClickListener = this

        credentialsLayout = view.findViewById(R.id.credentialsLayout)

        passwordLayout = view.findViewById(R.id.passwordLayout)
        passwordInput = view.findViewById(R.id.passwordEdit)
        passwordConfirm = view.findViewById(R.id.passwordConfirm)

        if (settings.blockAccessibility) {
            passwordLayout.importantForAccessibility = View.IMPORTANT_FOR_ACCESSIBILITY_NO_HIDE_DESCENDANTS
            passwordConfirm.importantForAccessibility = View.IMPORTANT_FOR_ACCESSIBILITY_NO
        }

        if (GeneralUtils.isOreo() && settings.blockAutofill) {
            passwordLayout.importantForAutofill = View.IMPORTANT_FOR_AUTOFILL_NO_EXCLUDE_DESCENDANTS
            passwordConfirm.importantForAutofill = View.IMPORTANT_FOR_AUTOFILL_NO
        }

        tooShortWarning = view.findViewById(R.id.tooShortWarning)

        passwordInput.addTextChangedListener(this)
        passwordConfirm.addTextChangedListener(this)

        val btnCancel = view.findViewById<Button>(R.id.btnCancel)
        btnSave = view.findViewById(R.id.btnSave)

        btnCancel.setOnClickListener(this)
        btnSave.setOnClickListener(this)

        updateLayout()

        super.onBindDialogView(view)
    }

    override fun onSetInitialValue(restorePersistedValue: Boolean, defaultValue: Any?) {
        if (restorePersistedValue) {
            val stringValue = getPersistedString(DEFAULT_VALUE.name.lowercase(Locale.ENGLISH))
            value = AuthMethod.valueOf(stringValue.uppercase(Locale.ENGLISH))
        } else {
            value = DEFAULT_VALUE
            persistString(value.name.lowercase(Locale.ENGLISH))
        }

        summary = entries[entryValues.indexOf(value)]
    }

    private fun saveValues() {
        if (settings.encryption == EncryptionType.PASSWORD) {
            if (value == AuthMethod.NONE || value == AuthMethod.DEVICE) {
                UIHelper.showGenericDialog(context, R.string.settings_dialog_title_error, R.string.settings_dialog_msg_auth_invalid_with_encryption)
                return
            }
        }

        if (value == AuthMethod.DEVICE) {
            val km = context.getSystemService(KEYGUARD_SERVICE) as KeyguardManager?

            if (!km!!.isKeyguardSecure) {
                Toast.makeText(context, R.string.settings_toast_auth_device_not_secure, Toast.LENGTH_LONG).show()
                return
            }
        }

        if (value == AuthMethod.PASSWORD || value == AuthMethod.PIN) {
            val password = passwordInput.text!!.toString()
            if (password.isEmpty())
                return

            // Nothing is stored yet. With password encryption the database key derives from the
            // credentials, so they may only replace the old ones once the database has been
            // re-encrypted with the new key; otherwise the old ones stay in effect.
            val newCredentials = settings.generateAuthCredentials(password) ?: return

            if (settings.encryption == EncryptionType.PASSWORD) {
                val callback = encryptionChangeCallback ?: return

                if (!callback.testEncryptionChange(newCredentials.key))
                    return
            }

            settings.saveAuthCredentials(newCredentials, value)
        }

        // Default-locale lowercase, as in the Java original (the key is read back with an
        // English-locale uppercase in Settings).
        persistString(value.toString().lowercase(Locale.getDefault()))
        summary = entries[entryValues.indexOf(value)]
    }

    override fun onClick(view: View) {
        // Resource ids are not compile-time constants any more (AGP 9), so no switch here.
        val id = view.id
        if (id == R.id.btnCancel) {
            dialog.dismiss()
        } else if (id == R.id.btnSave) {
            saveValues()
            dialog.dismiss()
        }
    }

    override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
        val password = passwordInput.editableText.toString()

        if (password.length >= minLength) {
            tooShortWarning.visibility = View.GONE

            val confirm = passwordConfirm.editableText.toString()

            btnSave.isEnabled = password.isNotEmpty() && confirm.isNotEmpty() && password == confirm
        } else {
            tooShortWarning.visibility = View.VISIBLE
        }
    }

    private fun updateLayout() {
        if (value == AuthMethod.NONE) {
            credentialsLayout.visibility = View.GONE

            if (dialog != null)
                UIHelper.hideKeyboard(context, dialog.currentFocus)

            btnSave.isEnabled = true
        } else if (value == AuthMethod.PASSWORD) {
            credentialsLayout.visibility = View.VISIBLE

            passwordLayout.hint = context.getString(R.string.settings_hint_password)
            passwordConfirm.setHint(R.string.settings_hint_password_confirm)

            passwordInput.inputType = InputType.TYPE_CLASS_TEXT or InputType.TYPE_TEXT_VARIATION_PASSWORD
            passwordConfirm.inputType = InputType.TYPE_CLASS_TEXT or InputType.TYPE_TEXT_VARIATION_PASSWORD

            passwordInput.transformationMethod = PasswordTransformationMethod()
            passwordConfirm.transformationMethod = PasswordTransformationMethod()

            minLength = Constants.AUTH_MIN_PASSWORD_LENGTH
            tooShortWarning.text = context.getString(R.string.settings_label_short_password, minLength)

            passwordInput.requestFocus()
            UIHelper.showKeyboard(context, passwordInput)

            btnSave.isEnabled = false
        } else if (value == AuthMethod.PIN) {
            credentialsLayout.visibility = View.VISIBLE

            passwordLayout.hint = context.getString(R.string.settings_hint_pin)
            passwordConfirm.setHint(R.string.settings_hint_pin_confirm)

            passwordInput.inputType = InputType.TYPE_CLASS_NUMBER or InputType.TYPE_NUMBER_VARIATION_PASSWORD
            passwordConfirm.inputType = InputType.TYPE_CLASS_NUMBER or InputType.TYPE_NUMBER_VARIATION_PASSWORD

            passwordInput.transformationMethod = PasswordTransformationMethod()
            passwordConfirm.transformationMethod = PasswordTransformationMethod()

            minLength = Constants.AUTH_MIN_PIN_LENGTH
            tooShortWarning.text = context.getString(R.string.settings_label_short_pin, minLength)

            passwordInput.requestFocus()
            UIHelper.showKeyboard(context, passwordInput)

            btnSave.isEnabled = false
        } else if (value == AuthMethod.DEVICE) {
            credentialsLayout.visibility = View.GONE

            if (dialog != null)
                UIHelper.hideKeyboard(context, dialog.currentFocus)

            btnSave.isEnabled = true
        }
    }

    override fun onItemClick(parent: AdapterView<*>?, view: View?, position: Int, id: Long) {
        value = entryValues[position]
        updateLayout()
    }

    // Needed stub functions
    override fun afterTextChanged(s: Editable?) {}

    override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {}

    companion object {
        @JvmField
        val DEFAULT_VALUE = AuthMethod.NONE

        private val entryValues = listOf(
            AuthMethod.NONE,
            AuthMethod.PASSWORD,
            AuthMethod.PIN,
            AuthMethod.DEVICE
        )
    }
}
